src/utils/tar.zig

```diff
@@ -0,0 +1,199 @@
+    const strip_components: u32 = 0;
+    loop: while (true) {
+        var header_buf = try allocator.alloc(u8, 512);
+        _ = try reader.readAtLeast(header_buf, 512);
+
+        const header: Header = .{ .bytes = header_buf[0..512] };
+
+        const file_size = try header.fileSize();
+        const rounded_file_size = std.mem.alignForward(u64, file_size, 512);
+        const pad_len = rounded_file_size - file_size;
+
+        var file_name_buf = try allocator.alloc(u8, 255);
+        const unstripped_file_name = try header.fullFileName(file_name_buf[0..255]);
+
+        switch (header.fileType()) {
+            .directory => {
+                const file_name = try stripComponents(unstripped_file_name, strip_components);
+                if (file_name.len != 0) {
+                    try dir.makePath(file_name);
+                }
+                allocator.free(header_buf);
+                allocator.free(file_name_buf);
+            },
+            .normal => {
+                if (file_size == 0 and unstripped_file_name.len == 0) {
+                    allocator.free(header_buf);
+                    allocator.free(file_name_buf);
+                    break :loop; // tar EOF
+                }
+
+                const file_name = try stripComponents(unstripped_file_name, strip_components);
+                if (std.fs.path.dirname(file_name)) |dir_name| {
+                    try dir.makePath(dir_name);
+                }
+
+                if (n_files_estimate) |n_files| {
+                    printTimeEstimate(&timer, n_files, file_count, "untar_files", null);
+                }
+                file_count += 1;
+
+                var contents = try allocator.alloc(u8, file_size);
+                _ = try reader.readAtLeast(contents, file_size);
+
+                try reader.skipBytes(pad_len, .{});
+
+                const entry = TarEntry{
+                    .contents = contents,
+                    .file_name = file_name,
+                    .filename_buf = file_name_buf,
+                    .header_buf = header_buf,
+                };
+
+                // find a free task
+                var task_ptr = &tasks[task_i];
+                while (!task_ptr.done.load(std.atomic.Ordering.Acquire)) {
+                    task_i = (task_i + 1) % n_threads;
+                    task_ptr = &tasks[task_i];
+                }
+                if (task_ptr.has_run) {
+                    allocator.free(task_ptr.entry.filename_buf);
+                    allocator.free(task_ptr.entry.header_buf);
+                    allocator.free(task_ptr.entry.contents);
+                    task_ptr.has_run = false;
+                }
+
+                task_ptr.entry = entry;
+                task_ptr.done.store(false, std.atomic.Ordering.Release);
+
+                const batch = Batch.from(&task_ptr.task);
+                thread_pool.schedule(batch);
+            },
+            .global_extended_header, .extended_header => {
```

Comment:
Looks like this is used for sparse files. Is this allowed or expected in a snapshot? If allowed, is skipping the bytes the correct way to handle it? If the answer to either of those questions is no, maybe we should log an error message if this is encountered.

Reply:
yeah great point - should be expected and i dont think this was the right way to handle it so changed it to throw an error